### PR TITLE
SNOW-1418523 make analyzer server connection thread safe

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1418523

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   **[NOTE] Easier to review by selecting `hide whitespace` option when viewing diffs** 
   This PR makes the following updates:
   1. `server_connection.py`: 
   - Adds `_thread_store` to create a `_cursor` per thread to enable concurrent querying
   - Protects `_query_listener` updates using `_lock` so all queries can be added in a thread-safe manner
   2. `session.py`: 
   - Create per thread `_analyzer` to enable concurrent dataframe operations.
   - Use lock to protect generation of actions ids
   - Use lock to protect running alter query tag and updating `_query_tag`.
   - Protect `_session_stage` creation with a lock so only one session stage is created and used.

### CODE REVIEW

**Comment by sfc-gh-aalam** [comment]
https://ci-dev-142.int.snowflakecomputing.com/job/PythonStoredProcBuildPrecommitTest/1322/

**Review by sfc-gh-yuwang — APPROVED** [review]

File: src/snowflake/snowpark/_internal/server_connection.py [hunk 1]
```diff
@@ -183,6 +185,12 @@ def __init__(
             "_skip_upload_on_content_match" in signature.parameters
         )
 
+    @property
+    def _cursor(self) -> SnowflakeCursor:
+        if not hasattr(self._thread_stored, "cursor"):
+            self._thread_stored.cursor = self._conn.cursor()
+        return self._thread_stored.cursor
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/server_connection.py:192]: does this operation need to be protected by self._lock?

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -623,8 +620,19 @@ def __str__(self):
...
+
+    @property
+    def _analyzer(self) -> Analyzer:
+        if not hasattr(self._thread_store, "analyzer"):
+            self._thread_store.analyzer = (
+                Analyzer(self)
+                if isinstance(self._conn, ServerConnection)
+                else MockAnalyzer(self)
+            )
+        return self._thread_store.analyzer
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/session.py:635]: same as above, do we need to protect this as well?

File: src/snowflake/snowpark/_internal/server_connection.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/server_connection.py:192]: this is a `threading.local()` object so each thread has its own copy. We don't need a lock for this step.

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:635]: same. this is a thread local object. this the how we create a per-thread instance of cursor and analyzer.

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**